File: Hashing/HashMap.java
package DSA;

public class HashMap {
	int size;
	String[] keys;
	String[] values;

	public HashMap(int size) {
		this.size = size;
		this.keys = new String[size];
		this.values = new String[size];
	}

	public int hashFunction(String key) {
		int sumOfChar = 0;
		for (char c : key.toCharArray()) {
			sumOfChar += (int) c;
		}
		return sumOfChar % size;
	}

	public void put(String key, String value) {
		int index = hashFunction(key);
		while (keys[index] != null && !keys[index].equals(key)) {
			index = (index + 1) % size;
		}
		keys[index] = key;
		values[index] = value;
	}

	String get(String key) {
		int index = hashFunction(key);
		while (keys[index] != null) {
			if (keys[index].equals(key))
				return values[index];
			index = (index + 1) & size;
		}

		return null;
	}

	void remove(String key) {
		int index = hashFunction(key);
		while (keys[index] != null) {
			if (keys[index].equals(key)) {
				keys[index] = null;
				values[index] = null;
				return;
			}
			index = (index + 1) % size;
		}
	}

	void display() {
		System.out.println("HashMap");
		for (int i = 0; i < size; i++) {
			if (keys[i] == null)
				System.out.println("Index " + i + ": " + "null");
			else
				System.out.println("Index " + i + ": " + keys[i] + " : " + values[i]);
		}
	}

	public static void main(String[] args) {
		HashMap hm = new HashMap(10);
		hm.put("James", "30");
		hm.put("Messi", "35");
		hm.put("Ronaldo", "38");
		hm.put("Neymar", "32");
		hm.display();
		System.out.print(hm.get("Messi"));
		hm.remove("Ronaldo");
		hm.display();
	}
}

File: Hashing/HashMap.py
class Hashmap:
    def __init__(self , size = 100):
        self.size = size 
        self.bucket = [[] for _ in range (size)]
    
    def hashFunction(self,key):
        return sum(ord(char) for char in key) % self.size
    def put(self, key , value ):
        index = self.hashFunction(key)
        bucket = self.bucket[index]
        for i , (k,v) in enumerate(bucket):
            if k == key :
                bucket[i] = (key , value)
                return 
        bucket.append((key , value))

    def get(self,key):
        index = self.hashFunction(key)
        bucket = self.bucket[index]
        for k , v in bucket:
            if k == key : 
                return v
        return None 
    
    def remove(self,key):
        index = self.hashFunction(key)
        bucket = self.bucket[index]
        for i , (k , v ) in enumerate(bucket):
            if k == key :
                del bucket[i]
                return 
    
    def display(self):
        print("Hash map")
        for index , bucket in enumerate(self.bucket):
            print(f"Bucket {index} : {bucket}")

if __name__ == "__main__" :
    hm = Hashmap(10)
    hm.display()
    #key is name , value is age 
    hm.put("James", 30)
    hm.put("Messi", 35)
    hm.put("Ronaldo", 38)
    hm.put("Neymar", 32)
    hm.display()

    print("Tuoi vua bong da la bong nhieu " , hm.get("Messi"))

    hm.remove("Ronaldo")
    hm.display()

